### ISSUE TITLE
Add Android and iOS support to Mesh Agent

### ISSUE THREAD

**Issue by whalehub** [Ylianst/MeshCentral#110]
I don't know if this is even technologically feasible with the way MeshCentral is built, but I was wondering if adding remote control capabilities for Android and iOS devices is something you are planning to implement in the future. What I mean by that is allowing people to install MeshAgent as an apk file or straight from Google Play Store on Android and then manage the device via MeshCentral.

**Comment by krayon007** [comment]
We previously had Android support, but had to remove it for several reasons. We used to use an undocumented API to capture the screen, but that required the device to be rooted. Also, use of undocumented APIs was blocked in recent versions of Android. There was a new API added to capture the screen, but it requires user action. Further, remote input is not supported with this API.

For iOS, we had issues with maintaining the control channel in the background. I was able to get it to work, but only by setting in the manifest that I was doing VoIP, which would most likely cause the app to be blocked from the App store.

**Comment by whalehub** [comment]
@krayon007 Thanks for your feedback. Do you know how TeamViewer manages to have remote control on unrooted Android devices? Are they using some sort of API to do it as well?

**Comment by krayon007** [comment]
When I used TeamViewer on my Android device, it was view only. (Remote input injection was unavailable). If this is how TeamViewer still works, it is most likely using the cast API which was added in either KitKat or Lollypop. I haven't used it in a while. Are you able to remotely interact with the device, or is it still view only?

**Comment by krayon007** [comment]
A quick Google search shows that remote input injection with TeamViewer requires an add-on pack to be installed separately that is device specific. That means it is most likely using an OEM sdk/interface to accomplish remote control.

**Comment by whalehub** [comment]
@krayon007 I was able to remotely interact with the device as well using one of the add-on packs you mentioned.

**Comment by whalehub** [comment]
@krayon007 Also, do you think you could give the iOS thing a try to see if the app will be blocked on the app store? I think it's worth a try.

**Comment by Ylianst** [comment]
So, there is no way we are going to be building a iOS or Android app at this point. Of course, if someone wants to work on this, it's an open source project and the help would be appreciated. However, this is not an option right now for us. We got plenty to do as it is and Android would be a big undertaking, IOS is likely not possible at all. I will close this issue. However, feel free to reopen if someone wants to work on this or continue the discussion.

**Comment by sourcex** [comment]
I have some heavy interest in this area (Android) and maybe having my team work on this.  Is the previous Android agent source available and a viable option for moving forward, or is this a greenfield project?

**Comment by jsastriawan** [comment]
Hi,

I am not sure if this is relevant. Android-x86, Chromeum OS x86 use Linux DRM API to render the graphic. I have attempted to use libDRM and uinput to substitute X11-based solution for Linux-based agent. If you want to have a look at it, please look at:

https://github.com/jsastriawan/MeshAgent

Obviously this will require root-ed Android/ChromiumOS.

For iOS, not sure if MeshAgent could get into Apple walled garden app store.
Regards,

Joko

**Comment by Patriciawa** [comment]
Maybe this is worth a shot? It works over TCP/IP, so you could make it work over port 443 right?
https://github.com/Genymobile/scrcpy

**Comment by MailYouLater** [comment]
If @Ylianst, @krayon007 or anyone else wants to look into this more, it would be really nice to have.
Anyway, I think STF might be similar enough that perhaps someone could figure out how their agent works, ... or maybe MeshCentral could 'just' be set up to accept connections from the STF agent?
https://openstf.io/
https://github.com/openstf/stf
https://github.com/openstf/STFService.apk

**Comment by amoljagdalepucsd** [comment]
Hi @krayon007 , you mentioned that we previously had Android support but had to remove it for several reasons. You also mentioned trying to maintain the control channel in the background for iOS by using VoIP. Were you able to manage the source code for these? If so, could you share it with us so we can take a look.
@Ylianst @jsastriawan @sourcex
